daemon/main.go

```diff
@@ -127,6 +132,22 @@ func main() {
 	}
 }
 
+func setupPidFile() {
+	pidfile.SetPidfilePath(defaults.PidFilePath)
+	if err := pidfile.Write(); err != nil {
+		log.WithError(err).Fatal("could not create pidfile")
+	}
+
+	// Handle the cleanup
+	sig := make(chan os.Signal, 1)
+	signal.Notify(sig, syscall.SIGQUIT, syscall.SIGINT, syscall.SIGHUP, syscall.SIGTERM)
```

Comment:
Oh, also, I think this `Notify` means the default handlers for these signals are no longer auto-exit. We'd need to handle them explicitly here, otherwise we can never stop the agent :/ I also wouldn't expect SIGHUP to be an exit signal unless it's left with the default handler (in my experience, it's used to reload configs but it's supposed to be a quit, I think).

Reply:
IIRC systemd is sending SIGHUP on stop or restart. Can you elaborate on what problem you are seeing? I am not following you.

Reply:
The SIGHUP thing is fine then, assuming we're always running under systemd.
The actual problem is that this is the only place we call `signal.Notify` and I think that suppresses any default handling of signals. The default would be to quit, I think. The goroutine we spawn here will read the channel, and keep removing the pidfile, but never quitting. This means we wouldn't actually exit when systemd sent us a QUIT or whatever.
I tried this with
```golang
package main

import (
        "fmt"
        "os"
        "os/signal"
        "syscall"
        "time"
)

func main() {
        sig := make(chan os.Signal, 1)
        signal.Notify(sig, syscall.SIGQUIT, syscall.SIGINT, syscall.SIGHUP, syscall.SIGTERM)
        go func() {
                for s := range sig {
                        fmt.Printf("saw %+v\n", s)
                }
        }()
        time.Sleep(30 * time.Second)
}
}
```

Reply:
@raybejjani Thanks for explaining. So is it enough to `os.Exit(0)` here or do you meaning something else by explicit?

Reply:
Would be great if you could create a issue for it. Thanks.

Reply:
yessir! https://github.com/cilium/cilium/issues/2063